ssl/quic/quic_port.c

```diff
@@ -1268,6 +1289,182 @@ static int port_validate_token(QUIC_PKT_HDR *hdr, QUIC_PORT *port,
+ * the address to packet header is kept in ptrs structure.
+ */
+static int port_chk_remove_hdr_protection(QUIC_PORT_SECRETS *qps,
+                                          QUIC_PKT_HDR_PTRS *ptrs)
+{
+    int l = 0;
+    int ok;
+    unsigned char pn_len, i;
+    unsigned char dst[16] = { 0 };
+    unsigned char mask[5] = { 0 };
+    unsigned char *first_byte;
+    unsigned char *pn_bytes;
+
+    l = 0;
+    ok = EVP_CipherUpdate(qps->qps_hpr_ctx, dst, &l, ptrs->raw_sample, 16);
+    if (ok == 0)
+        return 0;
+    /*
+     * We are going to drop header protection here. We decode packet number
+     * using the mask we got from raw_sample deciphered by header protector
+     * key.
+     */
+    for (i = 0; i < 5; ++i)
+         mask[i] = dst[i];
+    first_byte = ptrs->raw_start;
+    pn_bytes = ptrs->raw_pn;
+
+    *first_byte ^= mask[0] & ((*first_byte & 0x80) != 0 ? 0xf : 0x1f);
+    pn_len = (*first_byte & 0x3) + 1;
+    for (i = 0; i < pn_len; ++i)
+        pn_bytes[i] ^= mask[i + 1];
+
+    return 1;
+}
+
+static int port_chk_derive_nonce(QUIC_PORT_SECRETS *qps, QUIC_PN pn,
+                                 unsigned char *nonce, int nonce_sz)
+{
+    int nonce_len, i;
+
+    nonce_len = EVP_CIPHER_CTX_get_iv_length(qps->qps_pkt_ctx);
+    if (nonce_len > nonce_sz)
+        return 0;
+
+    memcpy(nonce, qps->qps_rx_cipher_iv, nonce_len);
+    for (i = 0; i < sizeof(QUIC_PN); ++i)
+        nonce[nonce_len - i - 1] ^= (unsigned char)(pn >> (i * 8));
+
+    return 1;
+}
+
+/*
+ * Function verifies integrity of INITIAL QUIC packet.
+ */
+static int port_chk_initial_pkt(QUIC_PORT *port, QUIC_URXE *e,
+                                QUIC_PKT_HDR *hdr, QUIC_PORT_SECRETS **pqps)
+{
+    PACKET pkt;
+    QUIC_PKT_HDR  hdr_local = { 0 };
+    QUIC_PKT_HDR_PTRS ptrs = { 0 };
+    unsigned char *work_buf = NULL;
+    size_t work_buf_sz = 0;
+    unsigned char scratch_dst[65535];
+    unsigned char nonce[EVP_MAX_IV_LENGTH];
+    const unsigned char *sop;
+    QUIC_PN pn;
+    size_t aad_len;
+    int l = 0, l2 = 0;
+    int ok;
+    QUIC_PORT_SECRETS *qps;
+
+    if (hdr->type != QUIC_PKT_TYPE_INITIAL)
+        return 0;
+
+    if (!PACKET_buf_init(&pkt, ossl_quic_urxe_data(e), e->data_len))
+        return 0;
+
+    if (pqps == NULL)
+        return 0;
+
+    /*
+     * create packet copy so qrx will obtain raw packet to do its own
+     * processing. This is sub-optimal. Ideally we want port to create qrx
+     * instance here to process packet.  if port will accept packet and create
+     * a channel later, then the qrx instance will be passed to channel
+     * constructor.
+     */
+    if (PACKET_memdup(&pkt, &work_buf, &work_buf_sz) == 0)
+        return 0;
+
+    if (!PACKET_buf_init(&pkt, work_buf, work_buf_sz)) {
+        return 0;
+    }
+    hdr = &hdr_local;
+    if (ossl_quic_wire_decode_pkt_hdr(&pkt, SIZE_MAX, 1, 0, hdr, &ptrs,
+                                      NULL) == 0)
+        return 0;
```

Comment:
I'm not sure if its worth doing, but we already parse the header in port_default_packet_header.  Would it be worth adding the ptrs parameter there, and then pass the hdr and ptrs parameters into this function, just to avoid the extra cycle?

Reply:
ok makes sense, we can walk before we run.